staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/immutable/algorithm.go

```diff
@@ -0,0 +1,550 @@
+				if !ok {
+					allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is immutable"))
+					continue
+				}
+
+				oldMap, ok := old[k].(map[string]interface{})
+				if !ok {
+					allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is immutable"))
+					continue
+				}
+
+				if !isImmutableMapKeysMutableValues(*prop.XKeyImmutability, oldMap, newMap) {
+					allErrs = append(allErrs, field.Invalid(pth.Child(k), newObj[k], "is immutable"))
+					continue
+				}
+				continue
+			}
+
+			// Non immutability, keep going deeper recursively
+			if prop.XImmutability == nil && prop.XKeyImmutability == nil {
+				if new == nil {
+					// new was deleted
+					continue
+				}
+				allErrs = append(allErrs, Immutable(old[k], newObj[k], &prop, pth.Child(k))...)
+				continue
+			}
+
+			continue
+		}
+	case []interface{}:
+		newArr, ok := new.([]interface{})
+		if !ok {
+			allErrs = append(allErrs, field.Invalid(pth.Child(s.Title), newArr, "is immutable"))
+			return allErrs
+		}
+		for k := range old {
+			allErrs = append(allErrs, Immutable(old[k], newArr[k], s.Items, nil)...)
+		}
+	default:
+		// scalars, do nothing
+	}
+	return allErrs
+}
+
+func isArrayMap(prop structuralschema.Structural) bool {
+	return prop.XListType != nil && *prop.XListType == xListTypeMap
+}
+
+func isSet(prop structuralschema.Structural) bool {
+	return prop.XListType != nil && *prop.XListType == xListTypeSet
+}
+
+// isMutableMapImmutableValues checks that values for existing keys are not mutated.
+func isMutableMapImmutableValues(oldMap, newMap map[string]interface{}) bool {
+	for key := range oldMap {
+		if _, ok := newMap[key]; ok {
+			if !equality.Semantic.DeepEqual(oldMap[key], newMap[key]) {
+				return false
+			}
+		}
+	}
+	return true
+}
+
+func derefPointerToString(pointer *string) string {
+	if pointer == nil {
+		return ""
+	}
+	return *pointer
+}
+
+// isMutableArrayImmutableItems
+// checks the items with their respective index in the array are immutable.
+// Hence, appending and removal at the end of the array are allowed, the change of an item or the change of the order are not.
+func isMutableArrayImmutableItems(old, new []interface{}) bool {
+	shortestArray := old
+	if len(new) < len(old) {
+		shortestArray = new
+	}
+	for key := range shortestArray {
+		if !equality.Semantic.DeepEqual(old[key], new[key]) {
+			return false
+		}
+	}
+	return true
+}
+
+func isMutableArrayMapImmutableItems(keys []string, old, new []interface{}) bool {
+	indexedList := make(map[string]interface{})
+	for i := range old {
+		oldItem, ok := old[i].(map[string]interface{})
+		if !ok {
+			return false
+		}
+
+		indexKey := ""
+		for _, key := range keys {
+			indexKey += fmt.Sprintf("%v", oldItem[key])
+		}
+		indexedList[indexKey] = oldItem
+	}
+
+	for i := range new {
+		newItem, ok := new[i].(map[string]interface{})
+		if !ok {
+			return false
+		}
+
+		indexKey := ""
+		for _, key := range keys {
+			indexKey += fmt.Sprintf("%v", newItem[key])
```

Comment:
are we sure this is a good encoding? I am not even sure that encoding is one-to-one, i.e. injective.

Reply:
Would rather use apimachinery json marshal.